### PR TITLE
Add support for loggers passed in as strings

### LINKED ISSUE #84: [feedback welcome] Support tensorboard, wandb loggers via strings
Tune currently has a variety of loggers supported: https://docs.ray.io/en/latest/tune/api_docs/logging.html#tbxlogger

We should allow this to also be accessible and configurable via tune-sklearn.

A proposal:

```
# Default to None
TuneSearchCV(loggers=None)

# Easy access
TuneSearchCV(loggers=["tensorboard"])

# support multiple loggers
TuneSearchCV(loggers=["tensorboard", "mlflow", "csv", "json"], logdir="./"))

# customize 
TuneSearchCV(loggers=["wandb"], logger_config={"..."})
```

### PR DESCRIPTION
**Pull request by inventormc**
Resolves #84

### CODE REVIEW

File: tune_sklearn/tune_basesearch.py [hunk 1]
```diff
@@ -253,7 +289,10 @@ def __init__(self,
                  return_train_score=False,
                  local_dir="~/ray_results",
                  max_iters=1,
-                 use_gpu=False):
+                 use_gpu=False,
+                 loggers=None,
+                 logdir="./",
+                 logconfig={}):
```
Comment (richardliaw) [inline tune_sklearn/tune_basesearch.py:295]: be careful - https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments

File: tune_sklearn/tune_search.py [hunk 2]
```diff
@@ -625,6 +632,7 @@ def _tune_run(self, config, resources_per_trial):
             config=config,
             fail_fast=True,
             resources_per_trial=resources_per_trial,
-            local_dir=os.path.expanduser(self.local_dir))
+            local_dir=os.path.expanduser(self.local_dir),
+            loggers=self.loggers)
```
Comment (richardliaw) [inline tune_sklearn/tune_search.py:636]: this should be a list of classes :)

File: tune_sklearn/tune_gridsearch.py [hunk 3]
```diff
@@ -116,21 +116,26 @@ class TuneGridSearchCV(TuneBaseSearchCV):
...
+            refit=True,
+            verbose=0,
+            error_score="raise",
+            return_train_score=False,
+            local_dir="~/ray_results",
+            max_iters=1,
+            use_gpu=False,
+            loggers=None,
+            logdir="./",
+            logconfig={},
```
Comment (richardliaw) [inline tune_sklearn/tune_gridsearch.py:137]: https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments !

File: tune_sklearn/tune_basesearch.py [hunk 4]
```diff
@@ -67,13 +69,47 @@ def resolve_early_stopping(early_stopping, max_iters):
...
+            elif log == "TBXLogger":
+                init_loggers.append(TBXLogger(logger_config, logdir))
+            elif log == "JsonLogger":
+                init_loggers.append(JsonLogger(logger_config, logdir))
+            elif log == "CSVLogger":
+                init_loggers.append(CSVLogger(logger_config, logdir))
+            elif log == "MLFLowLogger":
+                init_loggers.append(MLFLowLogger(logger_config, logdir))
+        elif isinstance(log, logger):
+            init_loggers.append(log)
```
Comment (richardliaw) [inline tune_sklearn/tune_basesearch.py:94]: what about just - "csv", "json", "tensorboard" - and automatically resolve to the logger class?

Comment (inventormc) [reply tune_sklearn/tune_basesearch.py]: Like take out all other loggers? Aren't these 3 the default loggers though

Comment (richardliaw) [reply tune_sklearn/tune_basesearch.py]: Sorry for the slow reply - basically, in comparison to Tune, let's do the following:

1. default to NO logger (tune.run(loggers=None))
2. User can do: `TuneSearchCV(loggers="csv")` or `TuneSearchCV(loggers=("csv", "tensorboard", "json"))`

if user passes in `"csv"`, you do:

```
init_loggers.append(CSVLogger)
```

Does that make sense?

Comment (inventormc) [reply tune_sklearn/tune_basesearch.py]: Yeah that makes sense. One thing: I thought json was required to have the analysis object we use to find the best metrics.

Comment (richardliaw) [reply tune_sklearn/tune_basesearch.py]: oh good point; guess it makes sense to default to always having json

File: tune_sklearn/tune_basesearch.py [hunk 5]
```diff
@@ -67,13 +70,50 @@ def resolve_early_stopping(early_stopping, max_iters):
...
+            if log == "UnifiedLogger":
+                init_loggers.append(UnifiedLogger)
+            elif log == "TBXLogger":
+                init_loggers.append(TBXLogger)
+            elif log == "JsonLogger":
+                init_loggers.append(JsonLogger)
+            elif log == "CSVLogger":
+                init_loggers.append(CSVLogger)
+            elif log == "MLFLowLogger":
+                init_loggers.append(MLFLowLogger)
```
Comment (richardliaw) [inline tune_sklearn/tune_basesearch.py:92]:
```suggestion
            if log == "tensorboard":
                init_loggers.append(TBXLogger)
            elif log == "json":
                init_loggers.append(JsonLogger)
            elif log == "csv":
                init_loggers.append(CSVLogger)
            elif log == "mlflow":
                init_loggers.append(MLFLowLogger)
```

File: tune_sklearn/tune_basesearch.py [hunk 6]
```diff
@@ -67,13 +70,50 @@ def resolve_early_stopping(early_stopping, max_iters):
...
 
 
+def resolve_loggers(loggers):
+    if loggers is None:
+        return None
+
+    if not isinstance(loggers, list):
+        raise TypeError("`loggers` must be a list of str or tune " "loggers.")
+
+    init_loggers = []
```
Comment (richardliaw) [inline tune_sklearn/tune_basesearch.py:80]: maybe use a set here instead, and pass into tune as a list

File: tune_sklearn/tune_basesearch.py [hunk 7]
```diff
@@ -67,13 +70,50 @@ def resolve_early_stopping(early_stopping, max_iters):
...
+                init_loggers.append(MLFLowLogger)
+        elif issubclass(log, Logger):
+            init_loggers.append(log)
+        else:
+            raise TypeError("`loggers` must be a list of str or tune "
+                            "loggers.")
+
+    for log in DEFAULT_LOGGERS:
+        if log not in init_loggers:
+            init_loggers.append(log)
```
Comment (richardliaw) [inline tune_sklearn/tune_basesearch.py:101]: if JsonLogger not in init_loggers:
    init_loggers.add(JsonLogger)

**Comment by richardliaw** [comment]
looks like there are merge conflicts?

File: tune_sklearn/tune_basesearch.py [hunk 8]
```diff
@@ -67,13 +70,45 @@ def resolve_early_stopping(early_stopping, max_iters):
                         f"or tune scheduler. Got {type(early_stopping)}.")
 
 
+def resolve_loggers(loggers):
+    if loggers is None:
+        return None
+
+    if not isinstance(loggers, list):
+        raise TypeError("`loggers` must be a list of str or tune " "loggers.")
```
Comment (richardliaw) [inline tune_sklearn/tune_basesearch.py:78]:
```suggestion
        raise TypeError("`loggers` must be a list of str or tune loggers.")
```

File: tune_sklearn/tune_gridsearch.py [hunk 9]
```diff
@@ -125,22 +125,26 @@ class TuneGridSearchCV(TuneBaseSearchCV):
...
-                 cv=5,
-                 refit=True,
-                 verbose=0,
-                 error_score="raise",
-                 return_train_score=False,
-                 local_dir="~/ray_results",
-                 max_iters=1,
-                 use_gpu=False,
-                 pipeline_auto_early_stop=True):
+<<<<<<< HEAD
```
Comment (richardliaw) [inline tune_sklearn/tune_gridsearch.py:128]: remove

File: tune_sklearn/tune_search.py [hunk 10]
```diff
@@ -334,6 +335,7 @@ def __init__(self,
             local_dir=local_dir,
             max_iters=max_iters,
             use_gpu=use_gpu,
+            loggers=loggers,
```
Comment (richardliaw) [inline tune_sklearn/tune_search.py:338]: docstring

File: tune_sklearn/tune_basesearch.py [hunk 11]
```diff
@@ -67,13 +70,45 @@ def resolve_early_stopping(early_stopping, max_iters):
...
+                init_loggers.add(TBXLogger)
+            elif log == "csv":
+                init_loggers.add(CSVLogger)
+            elif log == "mlflow":
+                init_loggers.add(MLFLowLogger)
+            elif log == "json":
+                init_loggers.add(JsonLogger)
+            else:
+                raise ValueError("{} is not a defined logger. "
+                                 "Check the list of available loggers."
```
Comment (richardliaw) [inline tune_sklearn/tune_basesearch.py:93]: Can you print the list of defined loggers? You could pass it in, or set it as a global variable

**Review by richardliaw — APPROVED** [review]

**Pull request merged**